scripts/cluster.py

```diff
@@ -54,270 +122,207 @@ def install_ray_over_ssh(node_ip_address, username, key_file, installation_direc
       ./install-dependencies.sh;
       ./setup.sh;
       ./build.sh
-    """.format(installation_directory, installation_directory)
-    run_command_over_ssh(node_ip_address, username, key_file, install_ray_command)
-  threads = []
-  for node_ip_address in node_ip_addresses:
-    t = threading.Thread(target=install_ray_over_ssh, args=(node_ip_address, username, key_file, installation_directory))
-    t.start()
-    threads.append(t)
-  for t in threads:
-    t.join()
-
-def _start_ray(node_ip_addresses, username, key_file, num_workers_per_node, worker_directory, installation_directory):
-  """
-  This method is used to start Ray on a cluster. It will ssh to the head node,
-    that is, the first node in the list node_ip_addresses, and it will start
-    the scheduler. Then it will ssh to each node and start an object store and
-    some workers.
-
-  :param node_ip_addresses: ip addresses of the nodes on which to install Ray
-  :param username: the username used to ssh to the cluster
-  :param key_file: the key used to ssh to the cluster
-  :param worker_directory: local directory containing the worker source code
-  :param installation_directory: directory in which Ray is installed, for example "/home/ubuntu/"
-  """
-  # First update the worker code on the nodes.
-  remote_worker_path = _update_worker_code(node_ip_addresses, worker_directory, installation_directory)
-
-  scripts_directory = os.path.join(installation_directory, "ray/scripts")
-  # Start the scheduler
-  # The triple backslashes are used for two rounds of escaping, something like \\\" -> \" -> "
-  start_scheduler_command = """
-    cd "{}";
-    source ../setup-env.sh;
-    python -c "import ray; ray.services.start_scheduler(\\\"{}:10001\\\", local=False)" > start_scheduler.out 2> start_scheduler.err < /dev/null &
-  """.format(scripts_directory, node_ip_addresses[0])
-  run_command_over_ssh(node_ip_addresses[0], username, key_file, start_scheduler_command)
-
-  # Start the workers on each node
-  # The triple backslashes are used for two rounds of escaping, something like \\\" -> \" -> "
-  for i, node_ip_address in enumerate(node_ip_addresses):
-    start_workers_command = """
+    """.format(self.installation_directory, self.installation_directory)
+    self._run_command_over_ssh_on_all_nodes_in_parallel(install_ray_command)
+
+  def start_ray(self, worker_directory, num_workers_per_node=10):
+    """Start Ray on a cluster.
+
+    This method is used to start Ray on a cluster. It will ssh to the head node,
+    that is, the first node in the list node_ip_addresses, and it will start the
+    scheduler. Then it will ssh to each node and start an object store and some
+    workers.
+
+    Args:
+      worker_directory (str): The path to the local directory containing the
```

Comment:
Here and in the other methods, we should describe a little better what this contains.